src/ocean/util/container/pool/AcquiredResources.d

```diff
@@ -0,0 +1,813 @@
+/*******************************************************************************
+
+    Helper structs for a specific execution context to acquire and relinquish
+    resources from a shared pool.
+
+    Several utilities are provided in this module, all of which build on top of
+    `FreeList`:
+        * `Acquired`: Tracks instances of a specific type acquired from the
+          shared resources and automatically relinquishes them when the
+          execution context exits.
+        * `AcquiredArraysOf`: Tracks arrays of a specific type acquired from the
+          shared resources and automatically relinquishes them when the
+          execution context exits.
+        * `AcquiredSingleton`: Tracks a singleton instance acquired from the
+          shared resources and automatically relinquishes it when the execution
+          context exits.
+
+    Examples of an "execution context" include `Task`s or connection handlers.
+
+    Copyright:
+        Copyright (c) 2016-2018 dunnhumby Germany GmbH.
+        All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module ocean.util.container.pool.AcquiredResources;
+
+import ocean.transition;
+import ocean.core.Verify;
+import ocean.util.container.pool.FreeList;
+
+/*******************************************************************************
+
+    Set of resources of the templated type acquired by an execution context. An
+    external source of elements of this type -- a FreeList!(T) -- as well as a
+    source of untyped buffers -- a FreeList!(ubyte[]) -- is required. When
+    resources are acquired (via the acquire() method), they are requested from
+    the free list and stored internally in an array. When the resources are no
+    longer required, the relinquishAll() method will return them to the free
+    list. Note that the array used to store the acquired resources is itself
+    acquired from the free list of untyped buffers and relinquished by
+    relinquishAll().
+
+    Params:
+        T = type of resource
+
+*******************************************************************************/
+
+public struct Acquired ( T )
+{
+    import ocean.util.container.VoidBufferAsArrayOf;
+
+    /// Type of a new resource.
+    static if ( is(typeof({T* t = new T;})) )
```

Comment:
Might be helpful to have a small comment to explain this `static if/else` choice ... ?  I presume it's meant to distinguish between cases where `T` is a value type and where `T` is a reference type?